File: src/contests/top150/n150_revaluate_reverse_polish_notation/Solution.java
package contests.top150.n150_revaluate_reverse_polish_notation;

import java.util.Stack;

public class Solution {

    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();

        for (String token: tokens) {
            switch (token) {
                case "+" -> stack.push(stack.pop() + stack.pop());
                case "*" -> stack.push(stack.pop() * stack.pop());
                case "-" -> {
                    int x = stack.pop();
                    int y = stack.pop();
                    stack.push(y - x);
                }
                case "/" -> {
                    int x = stack.pop();
                    int y = stack.pop();
                    stack.push(y / x);
                }
                default -> stack.push(Integer.valueOf(token));
            }
        }

        return stack.pop();
    }

    public static void main(String[] args) {
        var s = new Solution();
        System.out.println(s.evalRPN(new String[]{"1", "2", "+"}));
    }
}

File: src/contests/top150/n150_revaluate_reverse_polish_notation/solution.go
package main

import "strconv"

func evalRPN(tokens []string) int {
	var stack []int

	for i := 0; i < len(tokens); i++ {
		d, err := strconv.Atoi(tokens[i])
		if err == nil {
			stack = append(stack, d)
		} else {
			first := stack[len(stack)-2]
			second := stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			current := 0

			switch tokens[i] {
			case "+":
				current = first + second
			case "-":
				current = first - second
			case "*":
				current = first * second
			case "/":
				current = first / second
			}
			stack = append(stack, current)
		}
	}

	return stack[0]
}
